ktor-client/ktor-client-winhttp/windows/src/io/ktor/client/engine/winhttp/internal/WinHttpExceptions.kt

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2014-2022 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.client.engine.winhttp.internal
+
+import io.ktor.client.engine.winhttp.*
+import io.ktor.client.network.sockets.*
+import kotlinx.cinterop.*
+import platform.windows.*
+
+private val winHttpModuleHandle by lazy {
+    GetModuleHandleW("winhttp.dll")
+}
+private val languageId = MakeLanguageId(LANG_NEUTRAL, SUBLANG_DEFAULT)
+
+private val ERROR_INSUFFICIENT_BUFFER: UInt = platform.windows.ERROR_INSUFFICIENT_BUFFER.convert()
+private val ERROR_WINHTTP_TIMEOUT: UInt = platform.windows.ERROR_WINHTTP_TIMEOUT.convert()
+
+/**
+ * Creates an exception from last WinAPI error.
+ */
+internal fun getWinHttpException(message: String): Exception {
+    val errorCode = GetLastError()
+    return getWinHttpException(message, errorCode)
+}
+
+/**
+ * Creates an exception from WinAPI error code.
+ */
+internal fun getWinHttpException(message: String, errorCode: UInt): Exception {
+    val hResult = GetHResultFromError(errorCode)
+    val errorMessage = getErrorMessage(errorCode).trimEnd('.')
+    val cause = "$message: $errorMessage. Error $errorCode (0x${hResult.toString(16)})"
+
+    return if (errorCode == ERROR_WINHTTP_TIMEOUT) {
```

Comment:
I failed to find a way to distinguish which timeout may trigger this error during runtime and using `ConnectTimeoutException`. But maybe we need to use another exception class.

Reply:
It did actually the same things as Curl.
https://github.com/ktorio/ktor/blob/24c5d4654f6bdb68b28f9d4515953ef8b9d0b24b/ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlMultiApiHandler.kt#L293